File: PythonSpark/Utils/Writing.py
import logging

from pyspark.sql import DataFrame


def write_to_parquet(data: DataFrame, filepath: str) -> None:
    logging.info(f"Started writing data in parquet format to {filepath}")
    data.write.parquet(filepath)
    logging.info(f"Finished writing data in parquet format to {filepath}")


def write_to_postgres(data: DataFrame, url: str, user: str, password: str, table: str) -> None:
    logging.info(f"Started writing data to postgres with url {url} to a {table} table")
    data.write \
        .format("jdbc") \
        .option("url", url) \
        .option("user", user) \
        .option("password", password) \
        .option("dbtable", table) \
        .save()
    logging.info(f"Finished writing data to postgres with url {url} to a {table} table")


def write_to_big_query(data: DataFrame,
                       service_account_key_filepath: str,
                       temporary_bucket_name: str,
                       table: str) -> None:
    logging.info(f"Started writing data to BigQuery to a {table} table")
    data.write \
        .format("bigquery") \
        .option("temporaryGcsBucket", temporary_bucket_name) \
        .option("credentialsFile", service_account_key_filepath) \
        .save(table)
    logging.info(f"Finished writing data to BigQuery to a {table} table")

File: ScalaSpark/src/main/scala/Utils/Writing.scala
package Utils

import org.apache.spark.sql.{DataFrame, SaveMode}

import java.util.Properties

object Writing {
  def writeToParquet(data: DataFrame, filepath: String): Unit = {
    data.write.parquet(filepath)
  }

  def writeToPostgres(url: String, table: String, user: String, password: String, data: DataFrame): Unit = {
    val properties = new Properties()

    properties.put("user", user)
    properties.put("password", password)
    properties.put("driver", "org.postgresql.Driver")

    data.write
      .mode(SaveMode.Append)
      .jdbc(url, table, properties)
  }

  def writeToBigQuery(temporaryBucketName: String, table: String, keyFilepath: String, data: DataFrame): Unit = {
    data.write
      .format("bigquery")
      .option("temporaryGcsBucket", temporaryBucketName)
      .option("credentialsFile", keyFilepath)
      .save(table)
  }
}
